File: formcraft-plm/formcraft-plm/src/main/java/fr/formcraft/model/entity/CustomAttributeDefinition.java
package fr.formcraft.model.entity;

import fr.formcraft.model.enums.CustomAttributeType;
import fr.formcraft.model.enums.ProductType;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDateTime;

/**
 * Customization Gate 2 (BMIDE-style): a client-defined attribute that {@link Product}
 * instances may carry in their {@code customAttributes} map. Registering one of these
 * is a config-time action — no schema migration or Java code required.
 */
@Entity
@Table(name = "product_attribute_definitions")
@Getter
@Setter
@NoArgsConstructor
public class CustomAttributeDefinition {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "attribute_key", nullable = false, unique = true, length = 100)
    private String attributeKey;

    @Column(nullable = false, length = 150)
    private String label;

    @Enumerated(EnumType.STRING)
    @Column(name = "data_type", nullable = false, length = 20)
    private CustomAttributeType dataType;

    @Column(nullable = false)
    private boolean required;

    /** Restrict this attribute to one product type; null applies it to every product. */
    @Enumerated(EnumType.STRING)
    @Column(name = "applies_to_product_type", length = 30)
    private ProductType appliesToProductType;

    @Column(name = "validation_regex", length = 255)
    private String validationRegex;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt = LocalDateTime.now();
}

File: app/models/custom_attribute_definition.py
from __future__ import annotations

import datetime

from sqlalchemy import BigInteger, Boolean, DateTime, String
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base


class CustomAttributeDefinition(Base):
    """Customization Gate 2 (BMIDE-style): a client-defined attribute on Product. Registering
    one materialises a real column on products (see app.sdk.attributes) and records it in the
    customization repo's manifest — no Python code and no hand-written migration required."""

    __tablename__ = "product_attribute_definitions"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    attribute_key: Mapped[str] = mapped_column(String(100), unique=True, nullable=False)
    label: Mapped[str] = mapped_column(String(150), nullable=False)
    data_type: Mapped[str] = mapped_column(String(20), nullable=False)
    required: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    # Restrict this attribute to one product type; None applies it to every product.
    applies_to_product_type: Mapped[str | None] = mapped_column(String(30))
    validation_regex: Mapped[str | None] = mapped_column(String(255))
    # The physical products column this attribute was materialised as (x_<attribute_key>).
    column_name: Mapped[str | None] = mapped_column(String(63))
    created_at: Mapped[datetime.datetime] = mapped_column(
        DateTime, nullable=False, default=lambda: datetime.datetime.now(datetime.UTC)
    )
